ase/sw/ase_host_memory.c

```diff
@@ -0,0 +1,430 @@
+	int pt_level = ase_pt_length_to_level(length);
+	if (pt_level == -1)
+		return -1;
+
+	int status = ase_pt_pin_page(va, iova, pt_level);
+	if (status == 0) {
+		note_pinned_page(va, *iova, length);
+	}
+
+	pthread_mutex_unlock(&ase_pt_lock);
+	return status;
+}
+
+
+/*
+ * Unpin the page at iova.
+ */
+int ase_host_memory_unpin(uint64_t iova, uint64_t length)
+{
+	int status = 0;
+
+	pthread_mutex_lock(&ase_pt_lock);
+
+	if (ase_pt_root != NULL) {
+		int pt_level = ase_pt_length_to_level(length);
+		if (pt_level == -1)
+			return -1;
+
+		status = ase_pt_unpin_page(iova, pt_level);
+	}
+
+	pthread_mutex_unlock(&ase_pt_lock);
+
+	note_unpinned_page(iova, length);
+	return status;
+}
+
+
+/*
+ * Translate to simulated physical address space.
+ */
+uint64_t ase_host_memory_va_to_pa(void *va, uint64_t length)
+{
+	// We use a simple transformation to generate a fake physical space, merely
+	// inverting bits of the page address. Inverting these bits forces AFUs to
+	// translate addresses but avoids wasting simulator memory on complex page
+	// mappings.
+
+	// Confirm that the top virtual bits are 0. They should be non-zero only
+	// for kernel addresses.  This simulation is user space.
+	if ((uint64_t) va >> 48) {
+		ASE_ERR("Unexpected virtual address (0x" PRIx64 "), high bits set!", va);
+		raise(SIGABRT);
+	}
+
+	int pt_level = ase_pt_length_to_level(length);
+	uint64_t mask = 0xffffffffffffUL & (~0UL << (12 + 9 * pt_level));
+	return (uint64_t) va ^ mask;
+}
+
+/*
+ * Translate from simulated physical address space. Optionally hold the lock
+ * after successful translation so that the buffer remains pinned.
+ */
+void *ase_host_memory_pa_to_va(uint64_t pa, bool lock)
+{
+	// The inverse of ase_host_memory_va_to_pa.
+	if (pa >> 48) {
+		ASE_ERR("Unexpected physical address (0x" PRIx64 "), high bits set!", pa);
+		raise(SIGABRT);
+	}
+
+	// Is the page pinned?
+	int pt_level;
+	pthread_mutex_lock(&ase_pt_lock);
+	if (!ase_pt_check_addr(pa, &pt_level)) {
+		pthread_mutex_unlock(&ase_pt_lock);
+		return NULL;
+	}
+
+	if (!lock) {
+		pthread_mutex_unlock(&ase_pt_lock);
+	}
+
+	uint64_t mask = 0xffffffffffffUL & (~0UL << (12 + 9 * pt_level));
+	return (void *) (pa ^ mask);
+}
+
+
+void ase_host_memory_unlock(void)
+{
+	pthread_mutex_unlock(&ase_pt_lock);
+}
+
+
+void ase_host_memory_terminate(void)
+{
+	pthread_mutex_lock(&ase_pt_lock);
+
+	ase_pt_delete_tree(ase_pt_root, 3);
+	ase_pt_root = NULL;
+
+	pthread_mutex_unlock(&ase_pt_lock);
+}
+
+
+// ========================================================================
+//
+//	Maintain a simulated page table in order to track pinned pages.
+//
+// ========================================================================
+
+/*
+ * The page table here has a single job: indicate whether a given physical
+ * address (iova) is pinned. Since the simulator uses a simple XOR function
+ * for mapping virtual to physical addresses, the table is not needed for
```

Comment:
I like this simple hash function. But, consider the case where memory isn't backed by real hugepages (i.e. prepareBuffer uses sim_huge_mmap). What if the application spawns more than one process, each invoking fpgaPrepareBuffer separately? You could end up in a situation where identical VAs from both processes are mapped to the same PA. If the first process unpins the buffer, the buffer will be automatically unpinned for the second process right?

Reply:
Good observation. I thought about this too. The way ASE is currently structured I don't believe it could possibly support a pair of processes both talking to the RTL simulator. There are lots of statics and there is only one instance of channels to an application from the simulator. Updating the page mapping by adding an address allocator would be comparatively minor.

I think the address mapping is pretty well contained. We could switch to a range allocator and the largest change would be updating the ase_pt code in ase_host_memory to use the existing table to map pa to va. The memory footprint would grow, since the current table uses bit vectors (address present/not present) and the new table would need a full 64 bit va instead of a flag.

So, do we build it now or later if we go to multi-process? If now, we could commit the current code and I could add an allocator as a separate, smaller commit. What do you think?

Reply:
Multiprocess capability can be added later. Just wanted to highlight it here. I don't think it is a serious issue. I guess you could also create unique PAs hashing virtual address and process ID.